File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/pata/domain/PataObject.java
package com.zwstudio.logicpuzzlesandroid.puzzles.pata.domain;

public abstract class PataObject {
    public abstract String objTypeAsString();
    public static PataObject objTypeFromString(String str) {
        switch (str) {
            case "marker":
                return new PataMarkerObject();
            case "wall":
                return new PataWallObject();
            default:
                return new PataEmptyObject();
        }
    }
}

File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/pata/domain/PataObject.kt
package com.zwstudio.logicpuzzlesandroid.puzzles.pata.domainimport

import com.zwstudio.logicpuzzlesandroid.common.domain.Position
import com.zwstudio.logicpuzzlesandroid.home.domain.HintState

sealed class PataObject {
    open fun objTypeAsString() = "empty"

    companion object {
        fun objTypeFromString(str: String) = when (str) {
            "marker" -> PataMarkerObject()
            "wall" -> PataWallObject()
            else -> PataEmptyObject()
        }
    }
}

class PataEmptyObject : PataObject()

class PataHintObject(var state: HintState = HintState.Normal) : PataObject() {
    override fun objTypeAsString() = "hint"
}

class PataMarkerObject : PataObject() {
    override fun objTypeAsString() = "marker"
}

class PataWallObject(var state: HintState = HintState.Normal) : PataObject() {
    override fun objTypeAsString() = "wall"
}

class PataGameMove(val p: Position, var obj: PataObject = PataEmptyObject())
